support third party library configuration in qmake

tsteven4 opened the pull request:
The use of system libraries by distributions is common.  This is partially supported in the configure flow, but distributions usually require patches to achieve the linking they desire.

This PR gets us a step closer to deprecation/removal of the autotools base flow.

tsteven4 committed:
add qmake options to specify libraries.

tsteven4 committed:
fix no libusb along lines of #610

GPSBabelDeveloper commented on defs.h:
```diff
@@ -31,10 +31,16 @@
-#include <zlib.h>                 // doesn't really belong here, but is missing elsewhere.
-#elif !ZLIB_INHIBITED
-#include "zlib.h"                 // doesn't really belong here, but is missing elsewhere.
+#ifdef ZLIB_H_SYSTEM_INCLUDE
```
Are we slowly recreating C++17's _has_include? I think that's a feature that's actually been in most compilers before it was formalized.

tsteven4 replied:
I don't think so.  The build system may have both the system header and a copy in our source tree, and I want to respect the choice of the user instead of automatically deciding for them.

GPSBabelDeveloper commented on defs.h:
```diff
@@ -31,10 +31,16 @@
+#  include "zlib/zlib.h"              // doesn't really belong here, but is missing elsewhere.
+#else
+#  if HAVE_LIBZ
+#    include <zlib.h>                 // doesn't really belong here, but is missing elsewhere.
```
Insteead of changing all our sources to look in multiple places, have you considered using -I to add zlib and such to the list that's searched?
It's admittedly inefficient, but it may be less icky to debug the combinations in the field.

tsteven4 replied:
I intentionally didn't do that because I felt it was too easy too find the wrong zlib.h.  If we use project path when using the  library in our source tree it is unlikely we could find a system header file.   If we use brackets when using the system header it is unlikely we will find a header in our source.  I was trying to follow https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes which says "All of a project's header files should be listed as descendants of the project's source directory". 

The conditional includes could be simplified if we wanted to update configure and CMakeLists.  It could also be simplified like libusb if qmake didn't have a bug creating the vc template.

GPSBabelDeveloper commented on libusb.pri:
```diff
@@ -0,0 +1,64 @@
+  equals(WITH_LIBUSB, no) {
+    JEEPS += jeeps/gpsusbstub.cc
+  } else {
+    DEFINES += HAVE_LIBUSB_1_0
```
Is it time for HAVE_LIBUSB_1_0 to go? Isn't that effectively now HAVE_LIBUSB and we just have a floor of it being 1.0?

tsteven4 replied:
The library is actually named libusb-1.0.  So the convention used by AC_CHECK_LIB is to define HAVE_LIBUSB_1_0.  I intend to deprecate our use of autotools, but following the naming conventions they established seems better than creating our own.

GPSBabelDeveloper reviewed (approved):
I know a lot of this is just moving code around, but it feels like a lot of work to avoid using/duplicating such a small amount of code. Still, it's a good refactor, so go for it.
Let's do think if _has_include is our reason to move to C++17, though. (Looks like it's in GCC back to v5)
ISO Standing Document 6 has recommendations that help simplify code like this. Are you familiar with it?
If you want to commit this now as a checkpoint, that's OK with me. Marked approved.

tsteven4 committed:
correct inclusion of shapelib headers.

tsteven4 committed:
more closely match original includes

instead of trying to use HAVE_*, ENABLED_* *_INHIBITED and
-I options in a consistent manner.

We can revisit this after the demise of configure.

tsteven4 committed:
add qmake message when libusb is disabled.
